neuralmonkey/decoders/sequence_decoder.py

```diff
@@ -0,0 +1,310 @@
+    [("histories", Any),
+     ("constants", Any),
+     ("feedables", Any)])
+# pylint: enable=invalid-name
+
+# The LoopState is a structure that works with the tf.while_loop function
+# the decoder loop state stores all the information that is not invariant
+# for the decoder run.
+# pylint: disable=invalid-name
+DecoderHistories = NamedTuple(
+    "DecoderHistories",
+    [("logits", tf.TensorArray),
+     ("decoder_outputs", tf.TensorArray),
+     ("mask", tf.TensorArray)])  # float matrix, 0s and 1s
+
+DecoderConstants = NamedTuple(
+    "DecoderConstants",
+    [("train_inputs", Optional[tf.Tensor])])
+
+DecoderFeedables = NamedTuple(
+    "DecoderFeedables",
+    [("step", tf.Tensor),  # 1D int, number of the step
+     ("finished", tf.Tensor),  # batch-sized, bool
+     ("input_symbol", tf.Tensor),
+     ("prev_logits", tf.Tensor)])
+
+
+# pylint: disable=too-many-public-methods
+class SequenceDecoder(ModelPart):
+
+    def __init__(self,
+                 name: str,
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 max_output_len: int,
+                 dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        """Initialize parameters common for all autoregressive decoders.
+
+        Arguments:
+            name: Name of the decoder. Should be unique accross all Neural
+                Monkey objects.
+            vocabulary: Target vocabulary.
+            data_id: Target data series.
+            max_output_len: Maximum length of an output sequence.
+            dropout_keep_prob: Probability of keeping a value during dropout.
+        """
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+
+        log("Initializing decoder, name: '{}'".format(name))
+
+        self.vocabulary = vocabulary
+        self.data_id = data_id
+        self.max_output_len = max_output_len
+        self.dropout_keep_prob = dropout_keep_prob
+
+        # check the values of the parameters (max_output_len, ...)
+        if max_output_len <= 0:
+            raise ValueError("Maximum sequence length must be "
+                             "a positive integer.")
+
+        if dropout_keep_prob < 0.0 or dropout_keep_prob > 1.0:
+            raise ValueError("Dropout keep probability must be"
+                             "a real number in the interval [0,1].")
+
+        with self.use_scope():
+            self.train_mode = tf.placeholder(tf.bool, [], "train_mode")
+            self.go_symbols = tf.placeholder(tf.int32, [None], "go_symbols")
+
+            self.train_inputs = tf.placeholder(
+                tf.int32, [None, None], "train_inputs")
+            self.train_mask = tf.placeholder(
+                tf.float32, [None, None], "train_mask")
+
+    @tensor
+    def batch_size(self) -> tf.Tensor:
+        return tf.shape(self.go_symbols)[0]
+
+    @tensor
+    def decoding_w(self) -> tf.Variable:
```

Comment:
Docstring: projection matrix from decoder's output / hidden state to vocabulary. Klidně bych tam přidal note, že to často bývá největší matice modelu. Myslím, že nebude na škodu, když bude dokumentace zároveň tak trochu učebnice. Nebo je to blost?

Reply:
+1